src/Workspaces/Core/MSBuild.BuildHost/Rpc/RpcServer.cs

```diff
@@ -0,0 +1,175 @@
+{
+    private readonly TextWriter _sendingStream;
+    private readonly SemaphoreSlim _sendingStreamSemaphore = new SemaphoreSlim(initialCount: 1);
+    private readonly TextReader _receivingStream;
+
+    private readonly ConcurrentDictionary<int, object> _rpcTargets = new();
+    private volatile int _nextRpcTargetIndex = -1; // We'll start at -1 so the first value becomes zero
+
+    private readonly CancellationTokenSource _shutdownTokenSource = new CancellationTokenSource();
+
+    public RpcServer(Stream sendingStream, Stream receivingStream)
+    {
+        _sendingStream = new StreamWriter(sendingStream, JsonSettings.StreamEncoding);
+        _receivingStream = new StreamReader(receivingStream, JsonSettings.StreamEncoding);
+    }
+
+    public int AddTarget(object rpcTarget)
+    {
+        // Loop until we successfully have a new index for this; practically we don't expect this to ever collide, since that'd mean we'd have
+        // billions of long lived projects, but...
+        while (true)
+        {
+            var nextIndex = Interlocked.Increment(ref _nextRpcTargetIndex);
+            if (_rpcTargets.TryAdd(nextIndex, rpcTarget))
+                return nextIndex;
+        }
+    }
+
+    /// <summary>
+    /// Runs the server, waiting for responses. The task is completed when the receiving stream closes (and thus no more requests can come in), or
+    /// <see cref="Shutdown"/> is called.
+    /// </summary>
+    public async Task RunAsync()
+    {
+        var runningTasks = new ConcurrentSet<Task>();
+
+        string? line;
+        while ((line = await _receivingStream.TryReadLineOrReturnNullIfCancelledAsync(_shutdownTokenSource.Token).ConfigureAwait(false)) != null)
+        {
+            Request? request;
+
+            try
+            {
+                request = JsonConvert.DeserializeObject<Request>(line);
+                Contract.ThrowIfNull(request);
+            }
+            catch (Exception e)
+            {
+                throw new Exception($"Failure while deserializing '{line}'", innerException: e);
+            }
+
+            var runningTask = Task.Run(() => ProcessRequestAsync(request));
+
+            // We'll add this task to the list of running tasks, and then create a continuation to remove it from the list again; this ensures
+            // that we won't try to remove it before it was added in case the task completed by the time we got here.
+            runningTasks.Add(runningTask);
+            _ = runningTask.ContinueWith(
+                _ => Contract.ThrowIfFalse(runningTasks.Remove(runningTask)),
+                CancellationToken.None,
+                TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
+        }
+
+        // Wait until all outstanding requests are processed
+        await Task.WhenAll(runningTasks).ConfigureAwait(false);
+    }
+
+    private async Task ProcessRequestAsync(Request request)
+    {
+        Response response;
+
+        try
```

Comment:
my preference would a high level method that has the try/catch/finally, and almost nthing else.  and then have another method that is for the body of the try.

Reply:
Going to leave as is.